client/cypress/integration/login.spec.js

```diff
@@ -0,0 +1,123 @@
+/// <reference types="Cypress" />
+
+context('Login Test', () => {
+    before(() => {
+        cy.visit('');
+    });
+
+    beforeEach(() => {
+        cy.get('.login-container input[formcontrolname=username')
+            .as('user')
+            .parents('.mat-form-field')
+            .as('user-ff');
+        cy.get('.login-container input[formcontrolname=password')
+            .as('pw')
+            .parents('.mat-form-field')
+            .as('pw-ff');
+    });
+
+    it('redirects to login page', () => {
+        cy.location('pathname').should('eq', '/login');
+    });
+
+    it('has right title', () => {
+        cy.title().should('eq', 'Login - OpenSlides');
+    });
+
+    it('has fields & buttons', () => {
+        cy.get('@user').should('have.value', '');
+        cy.get('@pw').should('have.value', '');
+        cy.contains('.forgot-password-button', 'Forgot Password?').should('be.visible');
+        cy.get('.login-button').should('be.visible');
+    });
+
+    it('has right defaults', () => {
+        cy.get('@user').should('be.focused');
+        cy.get('@user-ff').should('not.have.class', 'mat-form-field-invalid');
+        cy.get('@pw').should('not.have.class', 'mat-form-field-invalid');
+        cy.get('@pw').should('have.attr', 'type', 'password');
+    });
+
+    it('adds error class on blur', () => {
+        cy.focused().blur();
+        cy.get('@user-ff').should('have.class', 'mat-form-field-invalid');
+        cy.get('@pw').focus();
+        cy.get('@pw-ff').should('not.have.class', 'mat-form-field-invalid');
+        cy.focused().blur();
+        cy.get('@pw-ff').should('have.class', 'mat-form-field-invalid');
+    });
+
+    it('shows error message', () => {
+        cy.get('.login-container .mat-error').should('not.be.visible');
+        cy.get('.login-button').click();
+        cy.get('.login-container .mat-error').should('have.text', 'Error: Username or password is not correct.');
+        cy.get('@user-ff').should('have.class', 'mat-form-field-invalid');
+        cy.get('@pw-ff').should('have.class', 'mat-form-field-invalid');
+    });
+
+    it('can make password visible', () => {
+        cy.get('@pw').should('have.attr', 'type', 'password');
+        cy.contains('.mat-icon', 'visibility').click();
+        cy.get('@pw').should('have.attr', 'type', 'text');
+    });
+
+    it('removes error classes', () => {
+        // test (hopefully) wrong user/pw combination
+        cy.get('@user').type(
+            '00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
+            { delay: 0 }
+        );
+        cy.get('@pw').type('wrong_password', { delay: 0 });
+        cy.get('@user-ff').should('not.have.class', 'mat-form-field-invalid');
+        cy.get('@pw-ff').should('not.have.class', 'mat-form-field-invalid');
+    });
+
+    it('rejects wrong login data', () => {
```

Comment:
Ok, one questions here: Do you have the server running while testing? If not you will get the same error message.

I strongly propose not to include the server in testing to seperate these two parts. We want to test the client not the server and sometimes we want to take control of responses (e.g. what happens, if the server sends garbage?) and the autoupdate system.

What is missing here is the capability to hook into requests and websocket-connections. I would suppose that a test must assert the correct data is send in a request and the test can specify the response, so it can be tested, if responses are handled correctly.

Another speciality from OpenSlides: The requests just initiate some changes. Responses are mostly empty and just contains a status. The new/changed/deleted data comes through an autoupdate.

So it would be nice, if we can do this in tests (example):
- I expect a request to /apps/users/login with data `{"username": "xx", "password": "yy"}`
- For this request I want to give `{"some_not_expected_data": "value"}` as a response with status code 400.
- Inject some (fake) autoupdate into the autoupdate service/websocket service.
- I expect the form to show error XXX, because of the response or the autoupdate data.

The autoupdate is more important for data-manipulating requests, e.g. to the rest-API. (not especially for login)